File: openAPI/src/test/java/jetbrains/exodus/env/EnvironmentConfigTest.java
/**
 * Copyright 2010 - 2016 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package jetbrains.exodus.env;

import jetbrains.exodus.ConfigSettingChangeListener;
import jetbrains.exodus.InvalidSettingException;
import jetbrains.exodus.core.dataStructures.hash.HashMap;
import org.jetbrains.annotations.NotNull;
import org.junit.Assert;
import org.junit.Test;

import java.util.Map;
import java.util.concurrent.atomic.AtomicBoolean;

public class EnvironmentConfigTest {

    @Test
    public void testCloneDefaults() {
        final Map<String, Object> defaults = EnvironmentConfig.DEFAULT.getSettings();
        final EnvironmentConfig ec = new EnvironmentConfig();
        final Map<String, String> stringDefaults = new HashMap<>();
        for (final Map.Entry<String, Object> entry : defaults.entrySet()) {
            stringDefaults.put(entry.getKey(), entry.getValue().toString());
        }
        ec.setSettings(stringDefaults);
        for (final Map.Entry<String, Object> entry : defaults.entrySet()) {
            Assert.assertEquals(entry.getValue(), ec.getSetting(entry.getKey()));
        }
    }

    @Test
    public void testListenerInvoked(){
        final Long oldValue = 1L;
        final Long newValue = 2L;
        final EnvironmentConfig ec = new EnvironmentConfig();
        final Boolean[] callBackMethodsCalled = new Boolean[2];
        ec.setMemoryUsage(oldValue);
        ec.addChangedSettingsListener(new ConfigSettingChangeListener() {
            @Override
            public void beforeSettingChanged(@NotNull String key, @NotNull Object value, @NotNull Map<String, Object> context) {
                Assert.assertEquals(EnvironmentConfig.MEMORY_USAGE, key);
                Assert.assertEquals(newValue, value);
                callBackMethodsCalled[0] = true;
            }

            @Override
            public void afterSettingChanged(@NotNull String key, @NotNull Object value, @NotNull Map<String, Object> context) {
                Assert.assertEquals(EnvironmentConfig.MEMORY_USAGE, key);
                Assert.assertEquals(newValue, value);
                callBackMethodsCalled[1] = true;
            }
        });

        ec.setMemoryUsage(newValue);

        Assert.assertArrayEquals(new Boolean[]{true, true}, callBackMethodsCalled);
    }

    @Test
    public void testListenerMethodsOrdering(){
        final Long oldValue = 1L;
        final Long newValue = 2L;
        final EnvironmentConfig ec = new EnvironmentConfig();
        ec.setMemoryUsage(oldValue);

        ec.addChangedSettingsListener(new ConfigSettingChangeListener() {
            @Override
            public void beforeSettingChanged(@NotNull String key, @NotNull Object value, @NotNull Map<String, Object> context) {
                Assert.assertEquals(oldValue, ec.getMemoryUsage());
            }

            @Override
            public void afterSettingChanged(@NotNull String key, @NotNull Object value, @NotNull Map<String, Object> context) {
                Assert.assertEquals(newValue, ec.getMemoryUsage());
            }
        });

        ec.setMemoryUsage(newValue);
    }

    @Test
    public void testListenerNotInvokedIfValueHasNotChanged(){
        final Long value = 1L;
        final EnvironmentConfig ec = new EnvironmentConfig();
        final Boolean[] callBackMethodsCalled = new Boolean[2];
        ec.setMemoryUsage(value);
        ec.addChangedSettingsListener(new ConfigSettingChangeListener() {
            @Override
            public void beforeSettingChanged(@NotNull String key, @NotNull Object value, @NotNull Map<String, Object> context) {
                callBackMethodsCalled[0] = true;
            }

            @Override
            public void afterSettingChanged(@NotNull String key, @NotNull Object value, @NotNull Map<String, Object> context) {
                callBackMethodsCalled[1] = true;
            }
        });

        ec.setMemoryUsage(value);

        Assert.assertArrayEquals(new Boolean[]{null, null}, callBackMethodsCalled);
    }

    @Test
    public void testListenerInvokedWithContext(){
        final Long oldValue = 1L;
        final Long newValue = 2L;
        final EnvironmentConfig ec = new EnvironmentConfig();
        ec.setMemoryUsage(oldValue);

        ec.addChangedSettingsListener(new ConfigSettingChangeListener() {

            private String contextKey = "key";
            private Object contextValue = new Object();

            @Override
            public void beforeSettingChanged(@NotNull String key, @NotNull Object value, @NotNull Map<String, Object> context) {
                Assert.assertTrue(context.isEmpty());
                context.put(contextKey, contextValue);
            }

            @Override
            public void afterSettingChanged(@NotNull String key, @NotNull Object value, @NotNull Map<String, Object> context) {
                Assert.assertEquals(contextValue, context.get(contextKey));
                Assert.assertEquals(1, context.size());
            }
        });

        ec.setMemoryUsage(newValue);
    }

    @Test
    public void testListenerCanBeSuppressed(){
        final Long oldValue = 1L;
        final Long newValue = 2L;
        final EnvironmentConfig ec = new EnvironmentConfig();
        final Boolean[] callBackMethodsCalled = new Boolean[2];
        ec.setMemoryUsage(oldValue);
        ec.addChangedSettingsListener(new ConfigSettingChangeListener() {
            @Override
            public void beforeSettingChanged(@NotNull String key, @NotNull Object value, @NotNull Map<String, Object> context) {
                callBackMethodsCalled[0] = true;
            }

            @Override
            public void afterSettingChanged(@NotNull String key, @NotNull Object value, @NotNull Map<String, Object> context) {
                callBackMethodsCalled[1] = true;
            }
        });

        EnvironmentConfig.suppressConfigChangeListenersForThread();
        ec.setMemoryUsage(newValue);
        EnvironmentConfig.resumeConfigChangeListenersForThread();

        Assert.assertArrayEquals(new Boolean[]{null, null}, callBackMethodsCalled);
    }

    @Test(expected = InvalidSettingException.class)
    public void testUnknownKey() {
        final Map<String, String> stringDefaults = new HashMap<>();
        stringDefaults.put("unknown.setting.key", null);
        new EnvironmentConfig().setSettings(stringDefaults);
    }
}

File: openAPI/src/test/kotlin/jetbrains/exodus/env/EnvironmentConfigTest.kt
/**
 * Copyright 2010 - 2018 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package jetbrains.exodus.env

import jetbrains.exodus.AbstractConfig
import jetbrains.exodus.ConfigSettingChangeListener
import jetbrains.exodus.InvalidSettingException
import jetbrains.exodus.core.dataStructures.hash.HashMap
import jetbrains.exodus.kotlin.notNull
import org.junit.Assert
import org.junit.Test

class EnvironmentConfigTest {

    @Test
    fun testCloneDefaults() {
        val defaults = EnvironmentConfig.DEFAULT.settings
        val ec = EnvironmentConfig()
        val stringDefaults = HashMap<String, String>()
        for ((key, value) in defaults) {
            stringDefaults.put(key, value.toString())
        }
        ec.setSettings(stringDefaults)
        for ((key, value) in defaults) {
            Assert.assertEquals(value, ec.getSetting(key))
        }
    }

    @Test
    fun testListenerInvoked() {
        val oldValue = 1L
        val newValue = 2L
        val ec = EnvironmentConfig()
        val callBackMethodsCalled = arrayOfNulls<Boolean>(2)
        ec.memoryUsage = oldValue
        ec.addChangedSettingsListener(object : ConfigSettingChangeListener {
            override fun beforeSettingChanged(key: String, value: Any, context: Map<String, Any>) {
                Assert.assertEquals(EnvironmentConfig.MEMORY_USAGE, key)
                Assert.assertEquals(newValue, value)
                callBackMethodsCalled[0] = true
            }

            override fun afterSettingChanged(key: String, value: Any, context: Map<String, Any>) {
                Assert.assertEquals(EnvironmentConfig.MEMORY_USAGE, key)
                Assert.assertEquals(newValue, value)
                callBackMethodsCalled[1] = true
            }
        })

        ec.memoryUsage = newValue

        Assert.assertArrayEquals(arrayOf(true, true), callBackMethodsCalled)
    }

    @Test
    fun testListenerMethodsOrdering() {
        val oldValue = 1L
        val newValue = 2L
        val ec = EnvironmentConfig()
        ec.memoryUsage = oldValue

        ec.addChangedSettingsListener(object : ConfigSettingChangeListener {
            override fun beforeSettingChanged(key: String, value: Any, context: Map<String, Any>) {
                Assert.assertEquals(oldValue, ec.memoryUsage)
            }

            override fun afterSettingChanged(key: String, value: Any, context: Map<String, Any>) {
                Assert.assertEquals(newValue, ec.memoryUsage)
            }
        })

        ec.memoryUsage = newValue
    }

    @Test
    fun testListenerNotInvokedIfValueHasNotChanged() {
        val value = 1L
        val ec = EnvironmentConfig()
        val callBackMethodsCalled = arrayOfNulls<Boolean>(2)
        ec.memoryUsage = value
        ec.addChangedSettingsListener(object : ConfigSettingChangeListener {
            override fun beforeSettingChanged(key: String, value: Any, context: Map<String, Any>) {
                callBackMethodsCalled[0] = true
            }

            override fun afterSettingChanged(key: String, value: Any, context: Map<String, Any>) {
                callBackMethodsCalled[1] = true
            }
        })

        ec.memoryUsage = value

        Assert.assertArrayEquals(arrayOf<Boolean?>(null, null), callBackMethodsCalled)
    }

    @Test
    fun testListenerInvokedWithContext() {
        val oldValue = 1L
        val newValue = 2L
        val ec = EnvironmentConfig()
        ec.memoryUsage = oldValue

        ec.addChangedSettingsListener(object : ConfigSettingChangeListener {

            private val contextKey = "key"
            private val contextValue = Any()

            override fun beforeSettingChanged(key: String, value: Any, context: MutableMap<String, Any>) {
                Assert.assertTrue(context.isEmpty())
                context.put(contextKey, contextValue)
            }

            override fun afterSettingChanged(key: String, value: Any, context: Map<String, Any>) {
                Assert.assertEquals(contextValue, context[contextKey])
                Assert.assertEquals(1, context.size.toLong())
            }
        })

        ec.memoryUsage = newValue
    }

    @Test
    fun testListenerCanBeSuppressed() {
        val oldValue = 1L
        val newValue = 2L
        val ec = EnvironmentConfig()
        val callBackMethodsCalled = arrayOfNulls<Boolean>(2)
        ec.memoryUsage = oldValue
        ec.addChangedSettingsListener(object : ConfigSettingChangeListener {
            override fun beforeSettingChanged(key: String, value: Any, context: Map<String, Any>) {
                callBackMethodsCalled[0] = true
            }

            override fun afterSettingChanged(key: String, value: Any, context: Map<String, Any>) {
                callBackMethodsCalled[1] = true
            }
        })

        EnvironmentConfig.suppressConfigChangeListenersForThread()
        ec.memoryUsage = newValue
        EnvironmentConfig.resumeConfigChangeListenersForThread()

        Assert.assertArrayEquals(arrayOf<Boolean?>(null, null), callBackMethodsCalled)
    }

    @Test(expected = InvalidSettingException::class)
    fun testUnknownKey() {
        val stringDefaults = HashMap<String, String>()
        stringDefaults.put("unknown.setting.key", null)
        EnvironmentConfig().setSettings(stringDefaults)
    }

    @Test
    fun suppressListenersInListener() {
        val settingFinished = arrayOf<Boolean?>(null)
        val ec = EnvironmentConfig()
        ec.addChangedSettingsListener(object : ConfigSettingChangeListener {
            override fun beforeSettingChanged(key: String, value: Any, context: Map<String, Any>) {
                settingFinished[0] = false
                AbstractConfig.suppressConfigChangeListenersForThread()
            }

            override fun afterSettingChanged(key: String, value: Any, context: Map<String, Any>) {
                settingFinished[0] = true
                AbstractConfig.resumeConfigChangeListenersForThread()
            }
        })
        ec.isGcEnabled = false
        val bool = settingFinished[0]
        Assert.assertNotNull(bool)
        Assert.assertTrue(bool.notNull)
    }
}
